### ISSUE TITLE
Empirical distribution draws are incorrect outside region of support

### ISSUE THREAD

**Issue by AaronDJohnson** [nanograv/enterprise_extensions#155]
`np.searchsorted` which is used to find the bin where the sample exists in the empirical distribution only gives the first or last bin number for a point outside the region of support for the empirical distribution. This means that the forward and reverse jump probabilities are wrong outside of the empirical distribution region, **breaking detailed balance**. I have a fix for this as suggested by @paulthebaker for `draw_from_empirical_distr` when `self.empirical_distr[distr_idx].ndim != 1` by setting `lQxy = -np.inf` when the point is outside support for the empirical distribution.

The same solution should work for the 1D case, but I need some sample data + an empirical distribution that doesn't cover the prior (we can change the prior range so any normal empirical distribution will work here).

`draw_from_psr_empirical_distr` may require a little more work than copy, paste. The same solution will probably work for this, but I haven't looked into it.

**Comment by meyers-academic** [comment]
@AaronDJohnson 

Maybe this is what you mean for the 1D case, but would it be simpler to set the bins a priori to cover the prior range? It looks like right now the bins are set to have edges at the min/max of the distribution [here](https://github.com/nanograv/enterprise_extensions/blob/master/enterprise_extensions/empirical_distr.py#L139). Would it be simple to just append the lower and upper end of the prior to those arrays? Or is the issue that we don't necessarily have access to the prior in that function?

Another option -- would using a 1D or 2D KDE work here as well? The `scipy.stats.gaussian_kde` seems to be very slow, but I have put together a simple version of `EmpiricalDistribution2D` that uses `sklearn.neighbors.KDE` and takes ~25 microseconds (which I think is still a good bit slower than the histogram method) and seems to give back sensible results. It might end up taking a bit less code to maintain, but I suppose would add a dependency.

```
class EmpiricalDistribution2D_alt(object):
    def __init__(self, param_names, samples):
        self.ndim = 2
        self.param_names = param_names

        # create a 2D KDE from which to evaluate
        self.kde = KernelDensity(kernel='gaussian', bandwidth=0.1).fit(samples)
    def draw(self):
        return self.kde.sample(1).T
    
    def prob(self, params):

        return np.exp(self.kde.score(np.array(params).reshape((1, 2))))

    def logprob(self, params):
        return self.kde.score(np.array(params).reshape((1, 2)))
```

**Comment by AaronDJohnson** [comment]
[Here](https://github.com/nanograv/enterprise_extensions/pull/156/commits/171b70435fcdd4dd3701ae00649cbe2e90f16d04) is the solution that I have ready at the moment for the 2D distributions. This is nice, because it won't require anyone to remake the empirical distribution if they decide to extend the prior domain. There is commented code that I _think_ will fix the 1D distributions, but I haven't tested that as of yet, because I don't have a sample with a 1D distribution.

I think these KDE distributions are worth a try! If they have no significant advantage in this context, I can think of another situation where they are likely to be useful.

**Comment by meyers-academic** [comment]
That makes sense for sure. Off the top of my head, I doubt they'd have a significant advantage in this context, other than that one would hope they avoid the issues that were just fixed. And if we're concerned about speed, as well, the KDE's took ~4x longer to draw from than drawing from the histograms.

Happy to try them in a different situation, though.

**Comment by Hazboun6** [comment]
@AaronDJohnson and @meyers-academic Can we close this issue now that both of your PRs have been merged?

**Comment by AaronDJohnson** [comment]
Yes! This issue has been fixed.
